fw/sock_clnt.c

```diff
@@ -342,12 +346,42 @@ tfw_h2_sk_prepare_xmit(struct sock *sk, struct sk_buff *skb,
 		 * We clear this flag to prevent it's copying
 		 * during skb splitting.
 		 */
-		if (!stream->xmit.b_len) {
+		if (!stream->xmit.t_len && !stream->xmit.b_len) {
 			skb_clear_tfw_flag(skb, SS_F_HTTT2_FRAME_DATA);
 			TFW_H2_STREAM_SEND_PROCESS(h2, stream, HTTP2_DATA);
 		}
 	}
 
+	if (FRAME_TRAILER_SHOULD_BE_MADE(flags)) {
+		r = tfw_h2_make_headers_frames(sk, skb, h2, stream, mss_now,
+					    *limit - stream->xmit.processed,
+					    &tmp_truesize);
+		if (unlikely(r)) {
+			T_WARN("%s: failed to make data frames (%d)",
+			       __func__, r);
+			if (r == -ENOMEM && headers_was_done) {
```

Comment:
Not necessary, because the skbs of trailers are separated from the DATA, so only after the skbs of HEADERS and DATA have been processed, it starts to do this part.